### DISCUSSION TITLE
Airflow db Migrate does not work

### DISCUSSION THREAD

**Discussion by aipatr** [apache/airflow#36328] [General] [answered]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### If "Other Airflow 2 version" selected, which one?

2.7.3

### What happened?

I try to run airflow db migrate from airflow 2.7.3 
and got the following

```console
/opt/miniconda3/envs/pytorch/lib/python3.10/site-packages/airflow/cli/cli_config.py:971 DeprecationWarning: The namespace option in [kubernetes] has been moved to the namespace option in [kubernetes_executor] - the old setting has been used, but please update your config.
DB: postgresql+psycopg2://airflow:***@localhost/airflow
Performing upgrade to the metadata database postgresql+psycopg2://airflow:***@localhost/airflow
[2023-12-20T10:43:58.244+0000] {migration.py:213} INFO - Context impl PostgresqlImpl.
[2023-12-20T10:43:58.245+0000] {migration.py:216} INFO - Will assume transactional DDL.
[2023-12-20T10:43:58.257+0000] {db.py:1620} INFO - Creating tables
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
Error in sys.excepthook:

Original exception was:
```

### How to reproduce

Installation of airflow 2.7.3
Postgres version psql (16.1 (Ubuntu 16.1-1.pgdg22.04+1), server 12.17 (Ubuntu 12.17-1.pgdg22.04+1))

### Operating System

Ubuntu 12.17

### Deployment

Virtualenv installation

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by Taragolis** [comment]
More like some troubleshooting rather than issue. 

Could you provide complete stack trace seems like your error is truncated here `Error in sys.excepthook:` and there `Original exception was:`

**Reply by aipatr** [reply]
How should I provide the stack trace? What I posted is what I see when running the command.

**Reply by Taragolis** [reply]
Is it anything more that just this lines in your output? Asked because according to this lines it is hard to say what is actually going wrong here

**Reply by Taragolis** [reply]
And how you upgrade your airflow installation? 

Did you follow [Installation from PyPI](https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html#)? E.g.: [Upgrading Airflow together with providers](https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html#upgrading-airflow-together-with-providers)

**Reply by aipatr** [reply]
I followed this https://airflow.apache.org/docs/apache-airflow/stable/installation/upgrading.html
I wanted to upgrade from 2.7.3 to 2.8
and that´s all the message that I got, otherwise I would have tried to solve it by myself. I completely agree that is obscure.

**Reply by potiuk** [reply]
You need to find out more information. Likely there is something in your system that hides the reminder of the stack trace - like redirecting of stderr for example somewhere else. Without we are unable to help you to solve your problem.

**Reply by potiuk** [reply]
Also - from the `excepthook` message it seems that somewhere in your deployment someone (maybe your admin) installed excepthook - to handle and redirect your exception and THAT's the reason you do not see anything. So you need ot rather ask whoever manages whatever deployment details you have - https://docs.python.org/3/library/sys.html#sys.excepthook and solve it with them. We are absolutely helpless here.

**Reply by aipatr** [reply]
I am the admin. The error happens in the shell when I run airflow db migrate so it should not have anything to do with Python sys.
the only pip package that has something related is exceptiongroup and airflow needs it. If I try to remove it I get : 
DB: postgresql+psycopg2://airflow:***@localhost/airflow
Performing upgrade to the metadata database postgresql+psycopg2://airflow:***@localhost/airflow
Traceback (most recent call last):
  File "/opt/miniconda3/envs/pytorch/bin/airflow", line 8, in <module>
    sys.exit(main())
  File "/opt/miniconda3/envs/pytorch/lib/python3.10/site-packages/airflow/__main__.py", line 57, in main
    args.func(args)
  File "/opt/miniconda3/envs/pytorch/lib/python3.10/site-packages/airflow/cli/cli_config.py", line 49, in command
    return func(*args, **kwargs)
  File "/opt/miniconda3/envs/pytorch/lib/python3.10/site-packages/airflow/utils/cli.py", line 114, in wrapper
    return f(*args, **kwargs)
  File "/opt/miniconda3/envs/pytorch/lib/python3.10/site-packages/airflow/utils/providers_configuration_loader.py", line 55, in wrapped_function
    return func(*args, **kwargs)
  File "/opt/miniconda3/envs/pytorch/lib/python3.10/site-packages/airflow/cli/commands/db_command.py", line 125, in migratedb
    db.upgradedb(
  File "/opt/miniconda3/envs/pytorch/lib/python3.10/site-packages/airflow/utils/session.py", line 79, in wrapper
    return func(*args, session=session, **kwargs)
  File "/opt/miniconda3/envs/pytorch/lib/python3.10/site-packages/airflow/utils/db.py", line 1576, in upgradedb
    import_all_models()
  File "/opt/miniconda3/envs/pytorch/lib/python3.10/site-packages/airflow/models/__init__.py", line 64, in import_all_models
    import airflow.models.serialized_dag
  File "/opt/miniconda3/envs/pytorch/lib/python3.10/site-packages/airflow/models/serialized_dag.py", line 35, in <module>
    from airflow.serialization.serialized_objects import DagDependency, SerializedDAG
  File "/opt/miniconda3/envs/pytorch/lib/python3.10/site-packages/airflow/serialization/serialized_objects.py", line 31, in <module>
    import cattr
  File "/opt/miniconda3/envs/pytorch/lib/python3.10/site-packages/cattr/__init__.py", line 1, in <module>
    from .converters import BaseConverter, Converter, GenConverter, UnstructureStrategy
  File "/opt/miniconda3/envs/pytorch/lib/python3.10/site-packages/cattr/converters.py", line 1, in <module>
    from cattrs.converters import (
  File "/opt/miniconda3/envs/pytorch/lib/python3.10/site-packages/cattrs/__init__.py", line 1, in <module>
    from .converters import BaseConverter, Converter, GenConverter, UnstructureStrategy
  File "/opt/miniconda3/envs/pytorch/lib/python3.10/site-packages/cattrs/converters.py", line 26, in <module>
    from ._compat import (
  File "/opt/miniconda3/envs/pytorch/lib/python3.10/site-packages/cattrs/_compat.py", line 66, in <module>
    from exceptiongroup import ExceptionGroup
ModuleNotFoundError: No module named 'exceptiongroup'

**Reply by potiuk** [reply]
You have not followed the instructions how to install/upgrade airflow. Airflow should be installed with constraints and neither conda nor miniconda does not support constraints. 

https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html

Remove old installation of airflow and minconda and install airflow from the scratch using the recommended method of installation we explained.

If you want to stick to miniconda, you will have to open the constraint file that you see in the instructions and manually synchronize all versions of dependencies that you will see in the right constraints file. But even then success is not quaranteed. So. Use pip and constraiints.

**Reply by aipatr** [reply]
I don´t know from what you got that I didn´t use the constraints. I did it for 2.7.3 using exactly the link that you are posting and I didn´t  have any issue. From what I see there is an issue in airflow, it is not reporting the exception and that´s not on my software if it happens when I run 
airflow db migrate

That is an airflow script

**Reply by aipatr** [reply]
@Taragolis  could you please help me? From other comments looks like it is always the fault of the poor user, once using exceptbook ( How can I influence an airflow script it is still a mistery to me). Another time guessing that I didn´t  use the guide, and again I don´t understand based on what. I want really resolve this issue and not simply uninstalling everything

**Reply by potiuk** [reply]
You have Airlfow installed in miniconda envronment. If you used both `pip` with constraints  and `conda` then it is even worse actually. Never use environment where you mix both `pip` and `conda`.

**Reply by potiuk** [reply]
Pip and conda are simply incompatible and lead to multiple strange issues if you mix them.

**Reply by aipatr** [reply]
I use pip and the environment is created using miniconda. What´s so strange about that? Is it a requirement of airflow that the environment needs to be created with everything except Conda or Miniconda? I installed airflow with PIP  and only pip.

**Reply by potiuk** [reply]
And just to add -> When you DID mix conda (in previously installed pytorch virtualenv - apparently installed with conda as is clearly visible in your stacktrace), removing it all and installing airlfow from scratch using new pip-only virtualenv will take you way less time than trying to fix your broken environment. You can keep the same database, and everything else, just start with a fresh virtualenv and pip-only as recommended by everyone in Python.

You might want to go and fix what's broken and spend more hours and days on fixing it, but when you do, I recommend you to come back here and see that you got a good advice.

**Reply by potiuk** [reply]
> I use pip and the environment is created using miniconda. What´s so strange about that? Is it a requirement of airflow that the environment needs to be created with everything except Conda or Miniconda? I installed airflow with PIP and only pip.

No idea. I just know it causes multiple issues that multiple users of ours had. But never had hours or days spare to investigate those.

**Reply by potiuk** [reply]
I just want to help you and save your time. Really.

**Reply by aipatr** [reply]
It is beautiful that instead of trying to understand where the db migrate is getting stuck and sending an exception, I am getting a lecture on something that YOU BELIEVE. If that´s how you want to treat your users I would strongly suggest that you focus on writing code more than answering to users that need help. Clearly, that´s not part of your expertise!

**Reply by aipatr** [reply]
Using Conda or Miniconda and then Pip is industry standard. If Airflow has an issue with it, it means that Airflow is not really stable.

**Reply by aipatr** [reply]
It does not sound at all like you want to help at all.

**Reply by potiuk** [reply]
I help as much as I can in my free time trying to help user like you to use less of their time on fighting with their problems. My default solution when your virtualenv gives you strange errors and you mix conda and pip is to do "clean" install. This usually solves problems with past installation in least amoutn of time (of both, the user who has problem and those who try to help them - again in their free time, away from their family, without the user paying for help, yet expecting that they will spend their time trying to understand all possible kinds of errors with the environment.

Let's see. But I expect that you will eventually do, what I suggested first after spending a lot of time on it. But I will shut up now, as apparenlty years of experience and multiple hundreds of users whom I helped (again for free, trying to put then on the fastest path to solve their problems) do not count here.

**Reply by Taragolis** [reply]
According to this output

```console
File "/opt/miniconda3/envs/pytorch/lib/python3.10/site-packages/airflow/models/serialized_dag.py", line 35, in
    from airflow.serialization.serialized_objects import DagDependency, SerializedDAG
File "/opt/miniconda3/envs/pytorch/lib/python3.10/site-packages/airflow/serialization/serialized_objects.py", line 31, in
    import cattr
```

There is something really bad with installation, [`cattrs`](https://pypi.org/project/cattrs/) completely removed from Airflow code codebase in favor of [`attrs`](https://pypi.org/project/attrs/) functionality, see: https://github.com/apache/airflow/blob/2.8.0/airflow/serialization/serialized_objects.py#L31

I would recommend install Airflow in completely new virtual environment and use same configuration and DB

**Reply by potiuk** [reply]
Yep. Same assesment here. the env is broken. You need a new one @aipatr

**Comment by aipatr** [comment]
Really helpful

Il 20 dic 2023 23:20, Jarek Potiuk ***@***.***> ha scritto:
You don't often get email from ***@***.*** Learn why this is important<https://aka.ms/LearnAboutSenderIdentification>

Yep. Same assesment here. the env is broken. You need a new one @aipatr<https://github.com/aipatr>

**Reply by potiuk** [reply]
> Really helpful

Helps to save you time. If you follow the advice that is. But if your point is to prove that airflow is unstable (despite having hundreds  of thousands users) and you are entitled to getting all the answer in the world and maintainers who try to help in their free time to get a crystal ball and figure out what environment you have and what you have done before with the environment and how you broke it, then your expectations are just unreasonable.

Maintainers here try to help if they can. They try to help you to solve it fast. Yes I uinderstand your ego got a bit hurt. Sorry for that apologies if by trying to tell you - yes briefly and succintly "install new environment, this way you will solve it the fastest possible way" your ego got hurt. Apologies. Se let me rephrase it. I hope it is going to be helpful this time:

We are really sorry to tell you that your environment is broken, unfortunately, the way it is broken is something that would require a lot of time to find out the reason for and both you and us will have to loose a lot of time on trying to find out the root cause. Therefore with a lot of regret, we are sorry to let you know that probably the best way for you to proceed is to recreate your virtual environment from scratch. It will likely take you minutes of doing it rather than hours or days for you and several maintainers assisiting you with your problem, trying to find out why you installed airflow in environment called `pytorch`, what other pckages you installed it, how miniconda interacted with your environment and what else you've done in the past to get the broken environment. We've alreaday lost a lot of time just getting to the point that your environment is broken and we think, that the assesment of time needed to reinstall the environment for you is far shorter. Also undoubtedly - as you - we have our families  that we want to meet in a few days for Holidays, so being away from them for too long is a bit painful, and if were to investigate it any longer, it would likely be affected, similarly the "reinstallation" advice we gave is likely one that will likely help you to also spend your time with your family for Holidays, and it will give you a peace of mind of problem being solved for you before. Undoubtedly you are emotionally connected to this problem, so it would likely disturb you during holidays.

I hope you will consider our kind advice and follow it, taking all that into account. And deep apologies if by being succint, non-emotional and expressing brevity, we hurt your ego. My intention for sure was to help you, if it came as condescending, then apologies. Advice is something you can take as present and open or you can throw it away - and unfortunately, similarly as I have no influence how you interpret the words written and how your feelings will be, I have controls over my intentions - which were clear - to help you to solve your problem fast.

It's entirely up to you what you do with such an advice. 

With best wishes for Holidays.

**Comment by aipatr** [comment, marked as answer]
For any other user that has the same problem, the solution is simple. No environment was broken, apparently some developers don´t take time to read text before answering.
If one want to upgrade using pip it should first
https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html

and only after the upgrade of the pip package
airflow db migrate

This is not transparent here. I write it for anyone else that could need this.

**Reply by potiuk** [reply]
>  If one want to upgrade using pip it should first
https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html

Apparently some users do not read what the maintainer write, because that was actually the FIRST advice you got from @taragolis

**Reply by potiuk** [reply]
And you told you've done that.

**Reply by aipatr** [reply]
before you entered in the discussion with your super long and useless messages, completely self absorbed on the fact that you know and others are stupid I wrote the following:
I followed this https://airflow.apache.org/docs/apache-airflow/stable/installation/upgrading.html
I wanted to upgrade from 2.7.3 to 2.8
and that´s all the message that I got, otherwise I would have tried to solve it by myself. I completely agree that is obscure.

In the link that I posted one could think that first of all one needs to run 
airflow db migrate 

and then upgrade 

but that´s not true. 
I am trying to help and clarify for other people that could have doubts. For you maybe it is super clear, reading the documentation it is not. If you want just to bully people on this platform I would strongly suggest that you don´t spend your precious time writing comments that are not helpful for anyone.

**Reply by potiuk** [reply]
How about this as a first advice you got:


> Did you follow [Installation from PyPI](https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html#)? E.g.: [Upgrading Airflow together with providers](https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html#upgrading-airflow-together-with-providers)


This is the first sugestion you get from @Taragolis . It contains

a) the link to the same instructions you marked as an aswer
b) the exact link on how exactly to upgrade airflow step by step


Did you miss it?

**Reply by potiuk** [reply]
Later in the discussion you also confirmed that you followed it.

Hence the conclusion that your environment is broken - because we assumed, you DID follow it. So I suggest to pay a bit more attention to the advises you get and to the answer you give. Because if you answer that you did something what you did not do, that might lead maintainers to wrong conclusions.

And it - unnecessarily - loses their time for needless discussions.

**Reply by potiuk** [reply]
And please, avoid throwing `bullying` accusations, where all what we want you to do is to help you and you are simply missing on following helpfull suggestions you got and set the trap on yourself but not following them.

**Reply by potiuk** [reply]
And just for the records. This is excerpt from the above discussion:

Jarek: 

>  You have not followed the instructions how to install/upgrade airflow. Airflow should be installed with constraints and neither conda nor miniconda does not support constraints.

https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html

You: 

> I don´t know from what you got that I didn´t use the constraints. I did it for 2.7.3 using exactly the link that you are posting and I didn´t have any issue

Then later:

You (answer):

> If one want to upgrade using pip it should first
https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html

> and only after the upgrade of the pip package
> airflow db migrate

If you still continue to claim that our answers were not helpful and that I was bullying you, then I think my brain explodes because we both with @Taragolis  advised you to do exactly what you finally did and you confirmed that you did it before. And only then we came to conclusion that your environment got broken.

**Comment by aipatr** [comment]
> And you told you've done that.
